hyperledger/sawtooth-core
# Implement PoET Fork Resolution and Refactoring Goodness

This update implements fork resolution using the same method previously used by PoET, with a few additions:

1.  If neither block is a PoET block, bail out.
2.  Otherwise, if one block is a PoET block and the other isn't, then the PoET block is chosen.
3.  Otherwise (both blocks are PoET blocks), if the blocks share the same immediate predecessor block, the one with the smaller wait duration is chosen.
4.  Otherwise (they have different predecessor blocks), the one with the greater aggregated local mean is chosen.
5.  Otherwise,  the one with the lexicographically greater block identifier (header signature) is chosen.

Criteria 1, 2, and 5 were added because prior to 0-8, non-PoET blocks were not a possibility.

In addition, the code has been refactored to better encapsulate the consensus state statistics and information used to make decision.

This PR updates:
* adds a fixed duration block count property to the PoET configuration view.  This value indicates the number of blocks which will have their local mean computed by using a simple ratio of target to initial wait times.  These blocks are not considered when performing the zTest.
* updates the PoET enclave factory to use the new PoET configuration view's fixed duration block count property.
* refactors existing PoET policy tests (key block claim limit, block claim delay) to common utils
* adds minimum win count and maximum win deviation properties to the PoET configuration view.  These values are used when performing the zTest.  The minimum win count is that minimum number of blocks, after the fixed duration block count has been surpassed, a validator has to win before zTest will apply to any further blocks.  The maximum win deviation is used to determine if a validator has violated the zTest policy by comparing this against the deviation from the expected number of blocks a validator should have won.
* adds aggregate local mean to the consensus state, properly updates it when blocks are committed, and updates fork resolution based upon steps outlined previously
* most of the code in the common utils has been refactored to live inside the consensus state class to better encapsulate it
* as a side effect of the refactoring, all consensus state properties are now read-only as the refactoring allowed for updating of properties to always take place inside of consensus state object methods

To test locally:
 `cd /project/sawtooth-core/integration/sawtooth_integration/docker` 

In `poet-smoke.yaml`, after the line:
`sawtooth.consensus.algorithm=poet \`

Add:

```
sawtooth.poet.target_wait_time=5 \
sawtooth.poet.initial_wait_time=0 \
sawtooth.poet.key_block_claim_count=5 \
sawtooth.poet.fixed_duration_block_count=10 \
```

Execute:
`run_docker_test poet-smoke.yaml`

Go get your beverage of choice and cross your fingers that the test passes.

After the test has completed, you should be able to locate in the log, messages from `poet_fork_resolver` indicating why a particular fork was chosen.
